utils/utils.go

```diff
@@ -165,6 +170,59 @@ func CleanUpIPAM(conf types.NetConf, args *skel.CmdArgs, logger *logrus.Entry) e
 	return err
 }
 
+// ReplaceHostLocalIPAMPodCIDRs extracts the host-local IPAM config section and replaces our special-case "usePodCidr"
+// subnet value with pod CIDR retrieved by the passed-in getPodCIDR function.  Typically, the passed-in function
+// would access the datastore to retrieve the podCIDR. However, for tear-down we use a dummy value that returns
+// 0.0.0.0/0.
+func ReplaceHostLocalIPAMPodCIDRs(logger *logrus.Entry, stdinData map[string]interface{}, getPodCIDR func() (string, error)) error {
+	ipamData := stdinData["ipam"].(map[string]interface{})
+	// Older versions of host-local IPAM store a single subnet in the top-level IPAM dict.
+	err := replaceHostLocalIPAMPodCIDR(logger, ipamData, getPodCIDR)
+	if err != nil {
+		return err
+	}
+	// Newer versions store one or more subnets in the "ranges" list:
+	untypedRanges := ipamData["ranges"]
+	if untypedRanges != nil {
+		rangeSets, ok := untypedRanges.([]interface{})
+		if !ok {
+			return fmt.Errorf("failed to parse host-local IPAM ranges section; was expecting a list, not: %v",
+				ipamData["ranges"])
+		}
+		for _, urs := range rangeSets {
+			rs, ok := urs.([]interface{})
+			if !ok {
+				return fmt.Errorf("failed to parse host-local IPAM range set; was expecting a list, not: %v", rs)
+			}
+			for _, r := range rs {
+				err := replaceHostLocalIPAMPodCIDR(logger, r, getPodCIDR)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func replaceHostLocalIPAMPodCIDR(logger *logrus.Entry, rawIpamData interface{}, getPodCidr func() (string, error)) error {
+	logrus.WithField("ipamData", rawIpamData).Debug("Examining IPAM data for usePodCidr")
+	ipamData := rawIpamData.(map[string]interface{})
```

Comment:
There _some_ logic.  Where I've used `_`, I'm handling the error case by checking for the zero value instead.  For example
```
subnet, _ := ipamData["subnet"].(string)
if subnet == ...
```
I.e. we only substitute the string "usePodCidr", otherwise we defer to the IPAM plugin to validate that field.

The unchecked cast of `stdinData["ipam"]` came from the existing code that I was updating but I might as well make that consistent.